quickwit/quickwit-indexing/src/actors/indexing_service.rs

```diff
@@ -332,6 +360,120 @@ impl IndexingService {
+            max_merge_factor: index_metadata
+                .indexing_settings
+                .merge_policy
+                .max_merge_factor,
+            split_num_docs_target: index_metadata.indexing_settings.split_num_docs_target,
+            ..Default::default()
+        };
+        let merge_policy: Arc<dyn MergePolicy> = Arc::new(stable_multitenant_merge_policy);
+
+        self.merge_policies.insert(key.clone(), merge_policy);
+        self.merge_policies.get(&key).unwrap().clone()
+    }
+
+    async fn get_or_init_indexing_directory(
+        &mut self,
+        pipeline_id: &IndexingPipelineId,
+        indexing_dir_path: PathBuf,
+    ) -> Result<IndexingDirectory, IndexingServiceError> {
+        let key = self.index_source_key(pipeline_id);
+        if let Some(indexing_directory) = self.indexing_directories.get(&key) {
+            return Ok(indexing_directory.clone());
+        }
+
+        let indexing_directory_path = indexing_dir_path
+            .join(&pipeline_id.index_id)
+            .join(&pipeline_id.source_id);
+        let indexing_directory = IndexingDirectory::create_in_dir(indexing_directory_path)
+            .await
+            .map_err(IndexingServiceError::InvalidParams)?;
+
+        self.indexing_directories
+            .insert(key.clone(), indexing_directory);
+        Ok(self.indexing_directories.get(&key).unwrap().clone())
+    }
+
+    fn get_or_init_split_store(
+        &mut self,
+        pipeline_id: &IndexingPipelineId,
+        cache_directory: &Path,
+        storage: Arc<dyn Storage>,
+        merge_policy: Arc<dyn MergePolicy>,
+    ) -> Result<IndexingSplitStore, IndexingServiceError> {
+        let key = self.index_source_key(pipeline_id);
+        if let Some(split_store) = self.split_stores.get(&key) {
+            return Ok(split_store.clone());
+        }
+
+        let split_store = IndexingSplitStore::create_with_local_store(
+            storage.clone(),
+            cache_directory,
+            IndexingSplitStoreParams {
+                max_num_bytes: self.split_store_max_num_bytes,
+                max_num_splits: self.split_store_max_num_splits,
+            },
+            merge_policy,
+        )?;
+
+        self.split_stores.insert(key.clone(), split_store);
+        Ok(self.split_stores.get(&key).unwrap().clone())
+    }
+
+    fn retain_merge_policies(&mut self, running_pipeline_ids: &[IndexingPipelineId]) {
+        let running_keys: HashSet<String> = running_pipeline_ids
+            .iter()
+            .map(|pipeline_id| pipeline_id.index_id.clone())
+            .collect();
+        let current_keys: HashSet<String> = self.merge_policies.keys().cloned().collect();
+        let deleted_keys = &current_keys - &running_keys;
+        for key in deleted_keys {
+            self.merge_policies.remove(&key);
+        }
+    }
+
+    fn retain_indexing_directories(&mut self, running_pipeline_ids: &[IndexingPipelineId]) {
```

Comment:
@guilload  `retain_indexing_directories` and `retain_split_stores` could be combined into one function but I kept them separate for simplicity. let me know if you feel it's ok to combine.

Reply:
That's fine as is.